server/common/app_config.py

```diff
@@ -19,74 +38,312 @@ def todict(self):
 
 
 class AppConfig(object):
-    def __init__(self, **kw):
-        super().__init__()
-
-        # app inputs
-        self.datapath = None
-        self.dataroot = None
-        self.title = ""
-        self.about = None
-        self.scripts = []
-        self.layout = []
-        self.max_category_items = 100
-        self.diffexp_lfc_cutoff = 0.01
-        self.disable_diffexp = False
-        self.enable_reembedding = False
-        self.anndata_backed = False
-
-        # The index page when in multi-dataset mode:
-        #   False or None:  this returns a 404 code
-        #   True:  loads a test index page, which links to the datasets that are available in the dataroot
-        #   string/URL:  redirect to this URL:  flask.redirect(config.multi_dataset_index)
-        self.multi_dataset_index = False
-
-        # A list of allowed matrix types.  If an empty list, then all matrix types are allowed
-        self.multi_dataset_allowed_matrix_type = []
-
-        # TODO these options may not apply to all datasets in the multi dataset.
-        # may need to invent a way to associate these config parameters with
-        # specific datasets.
-        self.obs_names = None
-        self.var_names = None
+    def __init__(self):
 
-        # parameters
-        self.diffexp_may_be_slow = False
-
-        inputs = [
-            "datapath",
-            "dataroot",
-            "title",
-            "about",
-            "scripts",
-            "layout",
-            "max_category_items",
-            "diffexp_lfc_cutoff",
-            "obs_names",
-            "var_names",
-            "anndata_backed",
-            "disable_diffexp",
-            "enable_reembedding",
-            "multi_dataset_index",
-            "multi_dataset_allowed_matrix_type",
-        ]
-
-        self.update(inputs, kw)
-
-    def update(self, inputs, kw):
-        for k, v in kw.items():
-            if k in inputs:
-                setattr(self, k, v)
-            else:
-                raise RuntimeError(f"unknown config parameter {k}.")
+        # load from default_config.yaml
+        default_config_file = pkg_resources.resource_filename(__name__, "default_config.yaml")
+        with open(default_config_file) as fyaml:
+            self.default_config = yaml.load(fyaml, Loader=yaml.FullLoader)
+
+        dc = self.default_config
+        self.server__verbose = dc["server"]["verbose"]
```

Comment:
these references into the default config will generate a KeyError if the default config is missing something or otherwise malformed.   You might want to wrap all of these references in a try/except, and throw a ConfigurationError if it is caught, indicating a bad install/packaging of the cellxgene module (ie, default_config.yaml is malformed)

Reply:
My thought was that we are in control of default_config.yaml, and if there is ever a KeyError, it would be a developer bug that we would fix.  But you are right, there could be an issue with an old or malformed default_config.   So, I think what I want to do is add the contents of "default_yaml.conif" to a python multiline string in app_config.py (or maybe a new file "default_yaml.py") and use that instead of opening and reading from a yaml file.  

1. this makes it so there is one less thing to worry about.  I can still add the try/except to give better error messages.

2. it solves my packaging problem (the reason why some of the tests failed to find default_config.yaml).  I won't have to have some special case for how to handle a non python file.  

3. I would like to add a new cli command: "cellxgene config" which returns the default config.   In this way a user can have a workflow where they ask cellxgene for the config, stash it in a config file, modify the file, then  launch cellxgene with their config file.

Reply:
that is a great solution.